src/core/qgsfeaturesource.h

```diff
@@ -85,6 +85,27 @@ class CORE_EXPORT QgsFeatureSource
      */
     virtual long featureCount() const = 0;
 
+    /**
+     * Determines if there are any features available in the source.
+     *
+     * In case it is not known if there are features available, false is returned.
+     * Use ``emptyUnknown()`` to determine if this value is reliable.
+     *
+     * \see QgsVectorDataProvider::empty() for data provider specific
+     *      information.
+     *
+     * \since QGIS 3.2
+     */
+    virtual bool empty() const;
+
+    /**
+     * Returns true if the value returned by ``empty()`` may be wrong.
+     * This depends on the dataprovider.
+     *
+     * \since QGIS 3.2
+     */
+    virtual bool emptyUnknown() const;
```

Comment:
I don't like this... (sorry to be blunt !) 

Could we instead remove this and do something like:
- check featureCount(), if !=-1 then we have the answer
- if -1, then do the feature request with a limit of 1 and no geometry and no attributes and see what we get?

Then we'd have a empty() method which can always be relied on without additional checks

Reply:
One of the main advantages of `empty()` over `featureCount()` is that it's fast. If the prerequisite to know if we can use the fast method is a call to a slow method that's rendering it a lot worse.

I wonder what should be done for a `QgsProcessingFeatureSource` for a filter algorithm. We don't know if it will be returning any features. And iterating over 1000000 features to find out it's empty is not ideal if it was just supposed to be a "quick check beforehand".

Reply:
Ok, then why not just fallback to the single feature request in the base method?

Reply:
Not sure I understand this. You mean requesting features with LIMIT 1 directly in the `empty()` method?

Reply:
Yeah, just like you've done with QgsVectorDataProvider::empty()

Reply:
As far as I know, processing feature sources do not (always) send their filter queries to the database, even less in future piped mode. In this case they'll locally evaluate expression or geometry filters.
If we want to know if a filter on a (potentially huge) source is empty we need to iterate and apply the expression to every feature which could lead to a costly loop. I want this method to be cheap by default. And I want to give the developer the required tools and knowledge to take an informed decision if getting reliable information might be more expensive.

Reply:
I'm a bit confused here - why would requesting one feature require us to iterate the whole layer? There's no query involved here as far as I can see.

Taking a few steps back, I wonder if an alternative approach would be to flip the method to "hasFeatures", and return the three state enum from that (where 0 = definitely no features)?

Reply:
> I'm a bit confused here - why would requesting one feature require us to iterate the whole layer? There's no query involved here as far as I can see.

If we are talking about (filter) algorithms (and not data providers), we'll have to run the algorithm until we find the first feature that passes the filter. If there are no features that pass the filter, we will have run the whole algorithm and consumed the whole data provider to find out that it's empty.

> Taking a few steps back, I wonder if an alternative approach would be to flip the method to "hasFeatures", and return the three state enum from that (where 0 = definitely no features)?

That sounds like an approach, I could live with that